flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -1115,6 +1117,15 @@ mlir::Value IntrinsicLibrary::genDim(mlir::Type resultType,
   return builder.create<mlir::SelectOp>(loc, cmp, diff, zero);
 }
 
+// DPROD
+mlir::Value IntrinsicLibrary::genDprod(mlir::Type resultType,
+				       llvm::ArrayRef<mlir::Value> args) {
+  assert(args.size() == 2);
+  assert(fir::isa_real(resultType) && (resultType.getKind() == 8) && "Result must be double precision in DPROD");
+  auto res = builder.create<mlir::MulFOp>(loc, args[0], args[1]);
+  return builder.createConvert(loc, resultType, res);
```

Comment:
From a standard point of view (Fortran 2018 16.9.64): "DPROD (X, Y) should have the same value as DBLE (X) * DBLE (Y).", so I think you should convert the arguments and to the result type before the multiplication rather than the result.

Reply:
Funnily enough, that's the first way I implemented it and I'm not sure why I changed it to the way it is now.  I'll change it back.